### DISCUSSION TITLE
Proposal: Class with conditional members

### DISCUSSION THREAD

**Discussion by webnickell** [dotnet/csharplang#3713] [Language Ideas]
Recently I come up with an idea, which I have never seen in another languages. What if classes can have conditional members? Like this: 
```cs
public class List<T>
    {
        //... some other methods

        if(T is int) T Sum()
        {
            var result = 0;
            foreach (var item in this)
                result += item;
            return result;
        }
    }
```
It means that method 'Sum' will be added only if type T is int or child of int. It will be added at compile time.  And if we use it, it will have such result
```cs
var intList = new List<int>();   //this list will have method Add
var stringList = new List<string>();    //but this no`
```
Also we could have support for duck typing system:
```cs
if (void Method() in T) T Sum() => 0;    //is true, if type T has method with such signature and name
if (string Property { get } in T) T Sum() => 0;   // is true, if type T has property with name, type and getter
if (string Field in T) T Sum() => 0;    //is true, if type T has field with such name and type
if (new(int) in T) T Sum() => 0;   //is true, if type T has ctor with such signature
```
Or more complex condition, like this:
```cs
if((T is int) && (void Method() in T)) T Sum() => 0;
```
Here we can check inheritance, binary operators (&& or ||) and duck typing system.
Maybe it would be useful for creating more flexible wrappers or containers

**Comment by huoyaoyuan** [comment]
> which I have never seen in another languages

You can get similar approach in C++ with template specialization.

> What if classes can have conditional members?

This isn't possible, because CLR doesn't allow generic specialization now.
There must be a **complete proposal** for CLR (Demonstrating how it works, not "I want this feature") before any language feature like this.

**Comment by CyrusNajmabadi** [comment]
`sum` can already be written as:

`public static int Sum(this IEnum<int> values) ...`

**Comment by webnickell** [comment]
>You can get similar approach in C++ with template specialization.

Yes, it is a little bit similar to C++ templates, but in C++ we can't check inheritance or existence of member. Also we can't simple add type members with conditional (maybe with some template magic).
>This isn't possible

Maybe it's possible via metadata for members, and this metadata would be used by JIT-compiler.
>sum can already be written as: public static int Sum(this IEnum<int> values) ...

Yes, but sometimes this way have some minuses. For example, we can't override this method for child collection (List, etc.). And if we write method 'Sum' for child type, we can't use it with IEnumerable param, cause compiler will always use static method.

**Comment by huoyaoyuan** [comment]
> Maybe it's possible via metadata for members, and this metadata would be used by JIT-compiler.

Personally, I do like this type of features. **But, there must be a complete CLR proposal (and then implementation).**

**Comment by PathogenDavid** [comment]
@webnickell
> but in C++ we can't check inheritance or existence of member.

You can, but like most things C++ metaprogramming [it's pretty gross](https://stackoverflow.com/questions/87372/check-if-a-class-has-a-member-function-of-a-given-signature).

> For example, we can't override this method for child collection (List, etc.). And if we write method 'Sum' for child type, we can't use it with IEnumerable param, cause compiler will always use static method.

It doesn't work when you want to add new types to a `Sum` method outside your control (although I don't think the original proposal covers that anyway), but Linq has [used this specialization approach for years](https://github.com/dotnet/runtime/blob/5c274f699d9e63d5e3054dff0b995ec186b96426/src/libraries/System.Linq/src/System/Linq/Count.cs#L18-L31) without issue.

This doesn't cover all of your scenarios, but it would cover some. Maybe shapes #164 covers the rest?

Besides shapes, there's also #1315

**Comment by GeraudFabien** [comment]
https://github.com/dotnet/csharplang/issues/3469
